lwgenerate/hcl.go

```diff
@@ -0,0 +1,270 @@
+	Name string
+
+	// Optional. Source for this provider
+	Source string
+
+	// Optional. Version for this provider
+	Version string
+}
+
+type HclProvider struct {
+	// Required, provider name
+	Name string
+
+	// Optional. Extra properties for this module.  Can supply string, bool, int, or map[string]interface{} as values
+	Attributes map[string]interface{}
+}
+
+type HclModule struct {
+	// Required, module name
+	Name string
+
+	// Required, source for this module
+	Source string
+
+	// Required, version
+	Version string
+
+	// Optional. Extra properties for this module.  Can supply string, bool, int, or map[string]interface{} as values
+	Attributes map[string]interface{}
+
+	// Optional.  Provider details to override defaults.  These values must be supplied as strings, and raw values will be
+	// accepted.  Unfortunately map[string]hcl.Traversal is not a format that is supported by hclwrite.SetAttributeValue
+	// today so we must work around it (https://github.com/hashicorp/hcl/issues/347).
+	ProviderDetails map[string]string
+}
+
+func convertTypeToCty(value interface{}) (cty.Value, error) {
+	switch v := value.(type) {
+	case string:
+		return cty.StringVal(v), nil
+	case int:
+		return cty.NumberIntVal(int64(v)), nil
+	case bool:
+		return cty.BoolVal(v), nil
+	default:
+		return cty.NilVal, errors.New("Unknown attribute value type")
+	}
+}
+
+// Helper to create various types of new hclwrite.Block using generic inputs
+func HclCreateGenericBlock(hcltype string, labels []string, attr map[string]interface{}) (*hclwrite.Block, error) {
+	block := hclwrite.NewBlock(hcltype, labels)
+
+	// Source and version require some special handling, should go at the top of a block declaration
+	sourceFound := false
+	versionFound := false
+
+	// We need/want to guarantee the ordering of the attributes, do that here
+	var keys []string
+	for k := range attr {
+		switch k {
+		case "source":
+			sourceFound = true
+		case "version":
+			versionFound = true
+		default:
+			keys = append(keys, k)
+		}
+	}
+	sort.Strings(keys)
+
+	if sourceFound || versionFound {
+		var newKeys []string
+		if sourceFound {
+			newKeys = append(newKeys, "source")
+		}
+		if versionFound {
+			newKeys = append(newKeys, "version")
+		}
+		keys = append(newKeys, keys...)
+	}
+
+	// Write block data
+	// TODO @ipcrm lists
+	for _, key := range keys {
+		val := attr[key]
+		switch v := val.(type) {
+		case string, int, bool:
+			value, err := convertTypeToCty(v)
+			if err != nil {
+				return nil, err
+			}
+			block.Body().SetAttributeValue(key, value)
+		case hcl.Traversal:
+			block.Body().SetAttributeTraversal(key, v)
+		case map[string]interface{}:
+			data := map[string]cty.Value{}
+			for attrKey, attrVal := range v {
+				value, err := convertTypeToCty(attrVal)
+				if err != nil {
+					return nil, err
+				}
+				data[attrKey] = value
+			}
+			block.Body().SetAttributeValue(key, cty.ObjectVal(data))
+		default:
+			return nil, errors.New("Unknown type")
+		}
+	}
```

Comment:
This block is a little difficult to read, perhaps extracting it to it's own function with a clear name and comments would help.

Reply:
👍